File: backend/utils/image.py
from PIL import Image
import io
from backend.constants import ALLOWED_EXTENSIONS


def allowed_file(filename):
    return '.' in filename and \
           filename.rsplit('.', 1)[1].lower() in ALLOWED_EXTENSIONS


def compress_image(image_data, max_size=(800, 800), quality=85):
    """Compress image to reduce file size while maintaining reasonable quality"""
    try:
        # Open image
        img = Image.open(io.BytesIO(image_data))

        # Convert to RGB if necessary
        if img.mode in ('RGBA', 'P'):
            img = img.convert('RGB')

        # Resize if too large
        img.thumbnail(max_size, Image.Resampling.LANCZOS)

        # Save compressed image
        output = io.BytesIO()
        img.save(output, format='JPEG', quality=quality, optimize=True)
        output.seek(0)

        return output.getvalue()
    except Exception as e:
        raise ValueError(f"Error processing image: {str(e)}")


def compress_banner_image(image_data, output_size=(1500, 300), quality=85):
    """
    Compress and center-crop image to banner aspect ratio (5:1).

    Takes the center portion of the image, crops to 5:1 aspect ratio,
    then resizes to the specified output size.

    Args:
        image_data: Binary image data
        output_size: Tuple of (width, height) for output, default (1500, 300)
        quality: JPEG quality 0-100, default 85

    Returns:
        Compressed JPEG binary data
    """
    try:
        img = Image.open(io.BytesIO(image_data))

        # Convert to RGB if necessary
        if img.mode in ('RGBA', 'P'):
            img = img.convert('RGB')

        # Calculate crop dimensions for 5:1 aspect ratio (center crop)
        target_ratio = output_size[0] / output_size[1]  # 5.0 for 1500x300
        img_ratio = img.width / img.height

        if img_ratio > target_ratio:
            # Image is wider than target - crop sides
            crop_height = img.height
            crop_width = int(crop_height * target_ratio)
            offset_x = (img.width - crop_width) // 2
            offset_y = 0
        else:
            # Image is taller than target - crop top/bottom
            crop_width = img.width
            crop_height = int(crop_width / target_ratio)
            offset_x = 0
            offset_y = (img.height - crop_height) // 2

        # Crop to target aspect ratio
        img = img.crop((offset_x, offset_y, offset_x + crop_width, offset_y + crop_height))

        # Resize to output size
        img = img.resize(output_size, Image.Resampling.LANCZOS)

        # Save compressed image
        output = io.BytesIO()
        img.save(output, format='JPEG', quality=quality, optimize=True)
        output.seek(0)

        return output.getvalue()
    except Exception as e:
        raise ValueError(f"Error processing banner image: {str(e)}")

File: frontend/src/utils/image.js
/**
 * Image Processing Utilities
 *
 * Shared utilities for image processing, cropping, and validation.
 * Used by profile picture, university logo, and banner upload components.
 */

/**
 * Configuration for square image processing (profile pictures, logos)
 */
export const IMAGE_CONFIG = {
  maxFileSize: 5 * 1024 * 1024, // 5MB
  outputSize: 400, // Output image dimensions (400x400)
  acceptedTypes: 'image/png,image/jpeg,image/jpg,image/gif,image/webp',
  quality: 0.9, // JPEG quality
};

/**
 * Configuration for banner image processing (5:1 aspect ratio)
 */
export const BANNER_CONFIG = {
  maxFileSize: 5 * 1024 * 1024, // 5MB
  outputWidth: 1500,
  outputHeight: 300,
  aspectRatio: 5, // width/height = 5:1
  acceptedTypes: 'image/png,image/jpeg,image/jpg,image/gif,image/webp',
  quality: 0.9,
};

/**
 * Center-crop an image to a square and return as Blob.
 *
 * Takes the center portion of the image, cropping to a square
 * based on the smaller dimension.
 *
 * @param {File} file - The image file to crop
 * @param {number} outputSize - The desired output dimensions (default: 400)
 * @param {number} quality - JPEG quality 0-1 (default: 0.9)
 * @returns {Promise<Blob>} The cropped image as a JPEG Blob
 *
 * @example
 * const croppedBlob = await cropImageToSquare(file);
 * await uploadProfilePicture(croppedBlob);
 */
export async function cropImageToSquare(
  file,
  outputSize = IMAGE_CONFIG.outputSize,
  quality = IMAGE_CONFIG.quality
) {
  return new Promise((resolve, reject) => {
    const img = new Image();
    const reader = new FileReader();

    reader.onload = (e) => {
      img.src = e.target.result;
    };

    img.onload = () => {
      try {
        const canvas = document.createElement('canvas');
        const ctx = canvas.getContext('2d');

        canvas.width = outputSize;
        canvas.height = outputSize;

        // Calculate crop dimensions (center crop to square)
        const size = Math.min(img.naturalWidth, img.naturalHeight);
        const offsetX = (img.naturalWidth - size) / 2;
        const offsetY = (img.naturalHeight - size) / 2;

        // Draw the center-cropped square
        ctx.drawImage(
          img,
          offsetX,
          offsetY,
          size,
          size,
          0,
          0,
          outputSize,
          outputSize
        );

        canvas.toBlob(
          (blob) => {
            if (blob) {
              resolve(blob);
            } else {
              reject(new Error('Failed to create image blob'));
            }
          },
          'image/jpeg',
          quality
        );
      } catch (error) {
        reject(error);
      }
    };

    img.onerror = () => {
      reject(new Error('Failed to load image'));
    };

    reader.onerror = () => {
      reject(new Error('Failed to read file'));
    };

    reader.readAsDataURL(file);
  });
}

/**
 * Validate an image file for upload.
 *
 * @param {File} file - The file to validate
 * @returns {{valid: boolean, error?: string}} Validation result
 */
export function validateImageFile(file) {
  if (!file) {
    return { valid: false, error: 'No file selected' };
  }

  if (file.size > IMAGE_CONFIG.maxFileSize) {
    return { valid: false, error: 'File size must be less than 5MB' };
  }

  if (!file.type.match(/^image\//)) {
    return { valid: false, error: 'Please select an image file' };
  }

  return { valid: true };
}

/**
 * Center-crop an image to banner aspect ratio (5:1) and return as Blob.
 *
 * Takes the center portion of the image, crops to 5:1 aspect ratio,
 * then scales to the specified output dimensions.
 *
 * @param {File} file - The image file to crop
 * @param {number} outputWidth - The desired output width (default: 1500)
 * @param {number} outputHeight - The desired output height (default: 300)
 * @param {number} quality - JPEG quality 0-1 (default: 0.9)
 * @returns {Promise<Blob>} The cropped image as a JPEG Blob
 *
 * @example
 * const croppedBlob = await cropImageToBanner(file);
 * await uploadBanner(croppedBlob);
 */
export async function cropImageToBanner(
  file,
  outputWidth = BANNER_CONFIG.outputWidth,
  outputHeight = BANNER_CONFIG.outputHeight,
  quality = BANNER_CONFIG.quality
) {
  return new Promise((resolve, reject) => {
    const img = new Image();
    const reader = new FileReader();

    reader.onload = (e) => {
      img.src = e.target.result;
    };

    img.onload = () => {
      try {
        const canvas = document.createElement('canvas');
        const ctx = canvas.getContext('2d');

        canvas.width = outputWidth;
        canvas.height = outputHeight;

        // Calculate crop dimensions for 5:1 aspect ratio (center crop)
        const targetRatio = outputWidth / outputHeight;
        const imgRatio = img.naturalWidth / img.naturalHeight;

        let cropWidth, cropHeight, offsetX, offsetY;

        if (imgRatio > targetRatio) {
          // Image is wider than target - crop sides
          cropHeight = img.naturalHeight;
          cropWidth = cropHeight * targetRatio;
          offsetX = (img.naturalWidth - cropWidth) / 2;
          offsetY = 0;
        } else {
          // Image is taller than target - crop top/bottom
          cropWidth = img.naturalWidth;
          cropHeight = cropWidth / targetRatio;
          offsetX = 0;
          offsetY = (img.naturalHeight - cropHeight) / 2;
        }

        // Draw the center-cropped region
        ctx.drawImage(
          img,
          offsetX,
          offsetY,
          cropWidth,
          cropHeight,
          0,
          0,
          outputWidth,
          outputHeight
        );

        canvas.toBlob(
          (blob) => {
            if (blob) {
              resolve(blob);
            } else {
              reject(new Error('Failed to create image blob'));
            }
          },
          'image/jpeg',
          quality
        );
      } catch (error) {
        reject(error);
      }
    };

    img.onerror = () => {
      reject(new Error('Failed to load image'));
    };

    reader.onerror = () => {
      reject(new Error('Failed to read file'));
    };

    reader.readAsDataURL(file);
  });
}
